### PR TITLE
feat(BMS): add datasource to query the list of BMS instances

### PR DESCRIPTION
**Pull request by MrLiangSir**
**What this PR does / why we need it**:
add datasource to query the list of BMS instances
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/bms TESTARGS='-run TestAccBmsInstancesDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/bms -v -run TestAccBmsInstancesDataSource_basic -timeout 360m -parallel 4
=== RUN   TestAccBmsInstancesDataSource_basic
=== PAUSE TestAccBmsInstancesDataSource_basic
=== CONT  TestAccBmsInstancesDataSource_basic
--- PASS: TestAccBmsInstancesDataSource_basic (550.62s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/bms       550.662s
```

### CODE REVIEW

File: docs/data-sources/bms_instances.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+* `region` - (Optional, String) Specifies the region in which to query the data source.
+  If omitted, the provider-level region will be used.
+
+* `enterprise_project_id` - (Optional, String) Specifies the enterprise project ID of BMS instance.
+
+* `flavor_id` - (Optional, String) Specifies the flavor ID of the desired flavor for the BMS instance.
+
+* `name` - (Optional, String) Specifies the name of the BMS instance.
+
+* `status` - (Optional, String) The status of the instance.
```
Comment (deer-hang) [inline docs/data-sources/bms_instances.md]:
```suggestion
* `status` - (Optional, String) Specifies the status of the instance.
```

File: docs/data-sources/bms_instances.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+* `flavor_id` - The ID of the BMS flavor.
+
+* `flavor_name` - The name of the BMS flavor.
+
+* `vcpus` - The number of vCPUs.
+
+* `memory` - The memory size in GB.
+
+* `disk` - The system disk size in the BMS flavor. The value `0` indicates that the disk size is not limited.
```
Comment (deer-hang) [inline docs/data-sources/bms_instances.md]: The unit of this field needs to be specified.

File: huaweicloud/services/acceptance/bms/data_source_huaweicloud_bms_instances_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		},
+	})
+}
+
+func testAccBmsInstancesDataSource_basic(name string) string {
+	return fmt.Sprintf(`
+%s
+
+data "huaweicloud_bms_instances" "test" {
+  name = huaweicloud_bms_instance.test.name
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/bms/data_source_huaweicloud_bms_instances_test.go]:
```suggestion
  depends_on = [huaweicloud_bms_instance.test]
```

File: huaweicloud/services/acceptance/bms/data_source_huaweicloud_bms_instances_test.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  name = data.huaweicloud_bms_instances.test.servers.0.name
+}
+output "name_filter_is_useful" {
+  value = length(data.huaweicloud_bms_instances.name_filter.servers) > 0 && alltrue(
+    [for v in data.huaweicloud_bms_instances.name_filter.servers[*].name : v == local.name]
+  )  
+}
+
+data "huaweicloud_bms_instances" "status_filter" {
+  name   = huaweicloud_bms_instance.test.name
```
Comment (deer-hang) [inline huaweicloud/services/acceptance/bms/data_source_huaweicloud_bms_instances_test.go]: Why do all filter conditions include name?

File: huaweicloud/services/bms/data_source_huaweicloud_bms_instances.go [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"agency_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"tags": {
+				Type:     schema.TypeList,
```
Comment (deer-hang) [inline huaweicloud/services/bms/data_source_huaweicloud_bms_instances.go]: The field type should be consistent with the resource and should be "map".

File: huaweicloud/services/bms/data_source_huaweicloud_bms_instances.go [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+	bmsClient, err := cfg.BmsV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating BMS client: %s", err)
+	}
+
+	listOpts := baremetalservers.ListOpts{
+		FlavorId:            d.Get("flavor_id").(string),
+		Name:                d.Get("name").(string),
+		Status:              d.Get("status").(string),
+		EnterpriseProjectId: cfg.GetEnterpriseProjectID(d),
```
Comment (deer-hang) [inline huaweicloud/services/bms/data_source_huaweicloud_bms_instances.go]:
```suggestion
		EnterpriseProjectId: cfg.DataGetEnterpriseProjectID(d),
```

**Comment by MrLiangSir** [comment]
make testacc TEST=./huaweicloud/services/acceptance/bms TESTARGS='-run TestAccBmsInstancesDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/bms -v -run TestAccBmsInstancesDataSource_basic -timeout 360m -parallel 4
=== RUN   TestAccBmsInstancesDataSource_basic
=== PAUSE TestAccBmsInstancesDataSource_basic
=== CONT  TestAccBmsInstancesDataSource_basic
--- PASS: TestAccBmsInstancesDataSource_basic (528.13s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/bms       528.170s

File: docs/data-sources/bms_instances.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+* `name` - (Optional, String) Specifies the name of the BMS instance.
+
+* `status` - (Optional, String) Specifies the status of the instance.
+  Value options are as follows:
+    + **ACTIVE**: running, stopping, deleting.
+    + **BUILD**: creating.
+    + **ERROR**: faulty.
+    + **HARD_REBOOT**: forcibly restarting.
+    + **REBOOT**: restarting.
+    + **SHUTOFF**: stopped, starting, deleting, rebuilding, reinstalling OS, OS reinstallation failed, frozen.
```
[the remark this answers is not in the document]
Comment (MrLiangSir) [reply docs/data-sources/bms_instances.md]: done

**Pull request merged**